File: doordeck-sdk/src/commonMain/kotlin/com/doordeck/multiplatform/sdk/model/Fusion.kt
package com.doordeck.multiplatform.sdk.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.js.JsExport

@JsExport
object Fusion {

    /**
     * The following classes are serializable because we are using them in the actual requests.
     */
    @Serializable
    sealed interface LockController

    @Serializable
    @SerialName("alpeta")
    data class AlpetaController(
        val username: String,
        val password: String,
        val doorId: Int, // Long, this can give problems
        val baseUrl: String? = null
    ) : LockController

    @Serializable
    @SerialName("amag")
    data class AmagController(
        val username: String,
        val password: String,
        val doorId: Int,
        val baseUrl: String? = null
    ) : LockController

    @Serializable
    @SerialName("assa-abloy")
    data class AssaAbloyController(
        val baseUrl: String,
        val doorId: String
    ) : LockController

    @Serializable
    @SerialName("avigilon")
    data class AvigilonController(
        val baseUrl: String,
        val username: String,
        val password: String,
        val doorId: String
    ) : LockController

    @Serializable
    @SerialName("axis")
    data class AxisController(
        val baseUrl: String,
        val doorIdentifier: String
    ) : LockController

    @Serializable
    @SerialName("demo")
    data class DemoController(
        val port: Int = 8080
    ) : LockController

    @Serializable
    @SerialName("gallagher")
    data class GallagherController(
        val baseUrl: String? = null,
        val apiKey: String,
        val doorId: String
    ) : LockController

    @Serializable
    @SerialName("genetec")
    data class GenetecController(
        val baseUrl: String,
        val username: String,
        val password: String,
        val doorId: String
    ) : LockController

    @Serializable
    @SerialName("lenel")
    data class LenelController(
        val baseUrl: String,
        val username: String,
        val password: String,
        val directoryId: String,
        val panelId: String,
        val readerId: String
    ) : LockController

    @Serializable
    @SerialName("mitrefinch")
    data class MitrefinchController(
        val host: String,
        val output: Int
    ) : LockController

    @Serializable
    @SerialName("net2")
    data class PaxtonNet2Controller(
        val host: String,
        val address: String,
        val output: Short
    ) : LockController

    @Serializable
    @SerialName("paxton10")
    data class Paxton10Controller(
        val baseUrl: String,
        val username: String,
        val password: String,
        val applianceId: Int
    ) : LockController

    @Serializable
    @SerialName("integra")
    data class IntegraV1Controller(
        val username: String,
        val password: String,
        val controllerId: Int
    ) : LockController

    @Serializable
    @SerialName("integra-v2")
    data class IntegraV2Controller(
        val baseUrl: String,
        val sessionId: String,
        val controllerId: Int,
        val cardholderId: Int,
        val pinCode: Int? = null
    ) : LockController

    @Serializable
    @SerialName("pac512")
    data class PacController(
        val dataSource: DataSource,
        val outputChannel: Int,
        val controllerSerial: Int
    ) : LockController

    @Serializable
    data class DataSource(
        val driverClass: String,
        val url: String,
        val user: String,
        val password: String
    )

    @Serializable
    @SerialName("tdsi-exgarde")
    data class TdsiExgardeController(
        val dbUrl: String? = null,
        val username: String,
        val password: String,
        val doorId: Int
    ) : LockController

    @Serializable
    @SerialName("tdsi-gardis")
    data class TdsiGardisController(
        val host: String,
        val username: String,
        val password: String,
        val doorId: Int
    ) : LockController

    @Serializable
    @SerialName("zkteco-zkbio-cvsecurity")
    data class ZktecoController(
        val clientSecret: String,
        val doorId: String,
        val baseUrl: String? = null,
        val entityType: ZktecoEntityType
    ) : LockController

    @Serializable
    enum class ZktecoEntityType {
        DOOR,
        FLOOR
    }
}

File: doordeck-sdk/src/mingwMain/resources/csharp/Model/Fusion.cs
using System.Net;
using System.Text.Json.Serialization;

namespace Doordeck.Headless.Sdk.Model;

public interface ILockController;

[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(AlpetaController), "alpeta")]
[JsonDerivedType(typeof(AmagController), "amag")]
[JsonDerivedType(typeof(AssaAbloyController), "assa-abloy")]
[JsonDerivedType(typeof(AvigilonController), "avigilon")]
[JsonDerivedType(typeof(AxisController), "axis")]
[JsonDerivedType(typeof(CCureController), "ccure")]
[JsonDerivedType(typeof(DemoController), "demo")]
[JsonDerivedType(typeof(GallagherController), "gallagher")]
[JsonDerivedType(typeof(GenetecController), "genetec")]
[JsonDerivedType(typeof(LenelController), "lenel")]
[JsonDerivedType(typeof(MitrefinchController), "mitrefinch")]
[JsonDerivedType(typeof(PaxtonNet2Controller), "net2")]
[JsonDerivedType(typeof(Paxton10Controller), "paxton10")]
[JsonDerivedType(typeof(IntegraV1Controller), "integra")]
[JsonDerivedType(typeof(IntegraV2Controller), "integra-v2")]
[JsonDerivedType(typeof(PacController), "pac512")]
[JsonDerivedType(typeof(TdsiExgardeController), "tdsi-exgarde")]
[JsonDerivedType(typeof(TdsiGardisController), "tdsi-gardis")]
[JsonDerivedType(typeof(ZktecoController), "zkteco-zkbio-cvsecurity")]
public abstract class LockController : ILockController;

public class AlpetaController : LockController
{
    public string? Username { get; set; }
    public string? Password { get; set; }
    public required int DoorId { get; set; }
    public Uri? BaseUrl { get; set; }
}

public class AmagController : LockController
{
    public required string Username { get; set; }
    public required string Password { get; set; }
    public required int DoorId { get; set; }
    public Uri? BaseUrl { get; set; }
}

public class AssaAbloyController : LockController
{
    public required Uri BaseUrl { get; set; }
    public required string DoorId { get; set; }
}

public class AvigilonController : LockController
{
    public required Uri BaseUrl { get; set; }
    public required string Username { get; set; }
    public required string Password { get; set; }
    public required string DoorId { get; set; }
}

public class AxisController : LockController
{
    public required Uri BaseUrl { get; set; }
    public required string DoorIdentifier { get; set; }
}

public class CCureController : LockController
{
    public Uri? BaseUrl { get; set; }
    public required string Username { get; set; }
    public required string Password { get; set; }
    public required string DoorType { get; set; }
    public required int DoorId { get; set; }
}

public class DemoController : LockController
{
    public required ushort Port { get; set; } = 8080;
}

public class GallagherController : LockController
{
    public Uri? BaseUrl { get; set; }
    public required string ApiKey { get; set; }
    public required string DoorId { get; set; }
}

public class GenetecController : LockController
{
    public required Uri BaseUrl { get; set; }
    public required string Username { get; set; }
    public required string Password { get; set; }
    public required string DoorId { get; set; }
}

public class LenelController : LockController
{
    public required Uri BaseUrl { get; set; }
    public required string Username { get; set; }
    public required string Password { get; set; }
    public required string DirectoryId { get; set; }
    public required string PanelId { get; set; }
    public required string ReaderId { get; set; }
}

public class MitrefinchController : LockController
{
    public required IPAddress Host { get; set; }
    public required int Output { get; set; }
}

public class PaxtonNet2Controller : LockController
{
    public required IPAddress Host { get; set; }
    public string? Username { get; set; }
    public string? Password { get; set; }
    public required string Address { get; set; }
    public required short Output { get; set; }
}

public class Paxton10Controller : LockController
{
    public required Uri BaseUrl { get; set; }
    public required string Username { get; set; }
    public required string Password { get; set; }
    public required int ApplianceId { get; set; }
}

public class IntegraV1Controller : LockController
{
    public required string Username { get; set; }
    public required string Password { get; set; }
    public required int ControllerId { get; set; }
}

public class IntegraV2Controller : LockController
{
    public required Uri BaseUrl { get; set; }
    public required string SessionId { get; set; }
    public required int ControllerId { get; set; }
    public required int CardholderId { get; set; }
    public int? PinCode { get; set; }
}

public class PacController : LockController
{
    public required DataSource DataSource { get; set; }
    public required int OutputChannel { get; set; }
    public required int ControllerSerial { get; set; }
}

public class DataSource
{
    public required string DriverClass { get; set; }
    public required string Url { get; set; }
    public required string User { get; set; }
    public required string Password { get; set; }
}

public class TdsiExgardeController : LockController
{
    public string? DbUrl { get; set; }
    public required string Username { get; set; }
    public required string Password { get; set; }
    public required int DoorId { get; set; }
}

public class TdsiGardisController : LockController
{
    public required IPAddress Host { get; set; }
    public required string Username { get; set; }
    public required string Password { get; set; }
    public required int DoorId { get; set; }
}

public class ZktecoController : LockController
{
    public required string ClientSecret { get; set; }
    public required string DoorId { get; set; }
    public string? BaseUrl { get; set; }
    public required ZktecoEntityType EntityType { get; set; }
}
    
[JsonConverter(typeof(JsonStringEnumConverter))]
public enum ZktecoEntityType
{
    DOOR,
    FLOOR
}